File: packages/stera_recorder/ios/stera_recorder/Sources/stera_recorder/coordination/RecordingLifecycleCoordinator.swift
import Foundation

class RecordingLifecycleCoordinator {

    struct PauseResult {
        let isPaused: Bool
        let pauseStartedAtMs: Int64
        let nextState: ArRecordingState
    }

    struct ResumeResult {
        let isPaused: Bool
        let pauseStartedAtMs: Int64
        let totalPausedDurationMs: Int64
        let nextState: ArRecordingState
    }

    struct CancelResult {
        let isRecording: Bool
        let isPaused: Bool
        let pauseStartedAtMs: Int64
        let totalPausedDurationMs: Int64
        let nextState: ArRecordingState
    }

    func pauseRecording(nowMs: Int64) -> PauseResult {
        return PauseResult(
            isPaused: true,
            pauseStartedAtMs: nowMs,
            nextState: .paused
        )
    }

    func resumeRecording(
        nowMs: Int64,
        pauseStartedAtMs: Int64,
        totalPausedDurationMs: Int64
    ) -> ResumeResult {
        let updatedPausedDurationMs: Int64
        if pauseStartedAtMs > 0 {
            updatedPausedDurationMs = totalPausedDurationMs + max(nowMs - pauseStartedAtMs, 0)
        } else {
            updatedPausedDurationMs = totalPausedDurationMs
        }
        return ResumeResult(
            isPaused: false,
            pauseStartedAtMs: 0,
            totalPausedDurationMs: updatedPausedDurationMs,
            nextState: .recording
        )
    }

    func cancelRecording() -> CancelResult {
        return CancelResult(
            isRecording: false,
            isPaused: false,
            pauseStartedAtMs: 0,
            totalPausedDurationMs: 0,
            nextState: .cancelling
        )
    }
}

File: packages/stera_recorder/android/src/main/kotlin/open/fpvlabs/stera/ar_recorder/coordination/RecordingLifecycleCoordinator.kt
package open.fpvlabs.stera.ar_recorder.coordination

import open.fpvlabs.stera.ar_recorder.ArRecordingState

class RecordingLifecycleCoordinator {
    data class PauseResult(
        val isPaused: Boolean,
        val pauseStartedAtMs: Long,
        val nextState: ArRecordingState
    )

    data class ResumeResult(
        val isPaused: Boolean,
        val pauseStartedAtMs: Long,
        val totalPausedDurationMs: Long,
        val nextState: ArRecordingState
    )

    data class CancelResult(
        val isRecording: Boolean,
        val isPaused: Boolean,
        val pauseStartedAtMs: Long,
        val totalPausedDurationMs: Long,
        val nextState: ArRecordingState
    )

    fun pauseRecording(nowMs: Long): PauseResult {
        return PauseResult(
            isPaused = true,
            pauseStartedAtMs = nowMs,
            nextState = ArRecordingState.PAUSED
        )
    }

    fun resumeRecording(
        nowMs: Long,
        pauseStartedAtMs: Long,
        totalPausedDurationMs: Long
    ): ResumeResult {
        val updatedPausedDurationMs = if (pauseStartedAtMs > 0L) {
            totalPausedDurationMs + (nowMs - pauseStartedAtMs).coerceAtLeast(0L)
        } else {
            totalPausedDurationMs
        }
        return ResumeResult(
            isPaused = false,
            pauseStartedAtMs = 0L,
            totalPausedDurationMs = updatedPausedDurationMs,
            nextState = ArRecordingState.RECORDING
        )
    }

    fun cancelRecording(): CancelResult {
        return CancelResult(
            isRecording = false,
            isPaused = false,
            pauseStartedAtMs = 0L,
            totalPausedDurationMs = 0L,
            nextState = ArRecordingState.CANCELLING
        )
    }
}
